ouroboros-network/src/Ouroboros/Network/KeepAlive.hs

```diff
@@ -8,43 +8,90 @@ module Ouroboros.Network.KeepAlive
   ( KeepAliveInterval (..)
   , keepAliveClient
   , keepAliveServer
+
+  , TraceKeepAliveClient
   ) where
 
-import           Control.Monad.Class.MonadSTM
-import           Control.Monad.Class.MonadTime (DiffTime)
+import           Control.Exception (assert)
+import qualified Control.Monad.Class.MonadSTM as Lazy
+import           Control.Monad.Class.MonadSTM.Strict
+import           Control.Monad.Class.MonadTime
 import           Control.Monad.Class.MonadTimer
+import           Control.Tracer (Tracer, traceWith)
+import qualified Data.Map.Strict as M
 
 import           Ouroboros.Network.Mux (RunOrStop (..), ScheduledStop)
+import           Ouroboros.Network.DeltaQ
 import           Ouroboros.Network.Protocol.KeepAlive.Client
 import           Ouroboros.Network.Protocol.KeepAlive.Server
 
 
 newtype KeepAliveInterval = KeepAliveInterval { keepAliveInterval :: DiffTime }
 
+data TraceKeepAliveClient peer =
+    AddSample peer DiffTime PeerGSV
+
+instance Show peer => Show (TraceKeepAliveClient peer) where
+    show (AddSample peer rtt gsv) = "AddSample " ++ show peer ++ " sample: " ++ show rtt
+        ++ " gsv: " ++ show gsv
 
 keepAliveClient
-    :: forall m.
+    :: forall m peer.
        ( MonadSTM   m
+       , MonadMonotonicTime m
        , MonadTimer m
+       , Ord peer
        )
-    => ScheduledStop m
+    => Tracer m (TraceKeepAliveClient peer)
+    -> ScheduledStop m
+    -> peer
+    -> (StrictTVar m (M.Map peer PeerGSV))
     -> KeepAliveInterval
+    -> StrictTVar m (Maybe Time)
```

Comment:
Do we need to access the above `StrictTVar` outside of the `keepAliveClient`?  If not I think we can remove it. The client can thread `Maybe Time` without using a `StrictTVar`.

Reply:
You're right it isn't needed.